pkg/reconciler/pubsub/topic/topic.go

```diff
@@ -0,0 +1,441 @@
+		// There was a difference and updateStatus did not return an error.
+		r.Recorder.Eventf(topic, corev1.EventTypeNormal, "Updated", "Updated Topic %q", topic.Name)
+	}
+	if reconcileErr != nil {
+		r.Recorder.Event(topic, corev1.EventTypeWarning, "InternalError", reconcileErr.Error())
+	}
+	return reconcileErr
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, topic *v1alpha1.Topic) error {
+	ctx = logging.WithLogger(ctx, r.Logger.With(zap.Any("topic", topic)))
+
+	topic.Status.InitializeConditions()
+
+	if topic.DeletionTimestamp != nil {
+		logging.FromContext(ctx).Desugar().Debug("Deleting Pub/Sub topic")
+		if topic.Spec.PropagationPolicy == v1alpha1.TopicPolicyCreateDelete {
+			if err := r.deleteTopic(ctx, topic); err != nil {
+				topic.Status.MarkNoTopic("TopicDeleteFailed", "Failed to delete Pub/Sub topic: %s", err.Error())
+				return err
+			}
+			topic.Status.MarkNoTopic("TopicDeleted", "Successfully deleted Pub/Sub topic: %s", topic.Status.TopicID)
+			topic.Status.TopicID = ""
+		}
+		removeFinalizer(topic)
+		return nil
+	}
+
+	// Add the finalizer.
+	addFinalizer(topic)
+
+	if err := r.reconcileTopic(ctx, topic); err != nil {
+		topic.Status.MarkNoTopic("TopicCreateFailed", "Failed to create Pub/Sub topic: %s", err.Error())
+		return err
+	}
+	topic.Status.MarkTopicReady()
+	// Set the topic being used.
+	topic.Status.TopicID = topic.Spec.Topic
+
+	err, svc := r.reconcilePublisher(ctx, topic)
+	if err != nil {
+		topic.Status.MarkNotDeployed("PublisherCreateFailed", "Failed to create Publisher: %s", err.Error())
+		return err
+	}
+
+	// Update the topic.
+	topic.Status.PropagatePublisherStatus(svc.Status.GetCondition(apis.ConditionReady))
+	if svc.Status.IsReady() {
+		topic.Status.SetAddress(svc.Status.Address.URL)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) reconcileTopic(ctx context.Context, topic *v1alpha1.Topic) error {
+	if topic.Status.ProjectID == "" {
+		projectID, err := utils.ProjectID(topic.Spec.Project)
+		if err != nil {
+			logging.FromContext(ctx).Desugar().Error("Failed to find project id", zap.Error(err))
+			return err
+		}
+		// Set the projectID in the status.
+		topic.Status.ProjectID = projectID
+	}
+
+	// Auth to GCP is handled by having the GOOGLE_APPLICATION_CREDENTIALS environment variable
+	// pointing at a credential file.
+	client, err := r.createClientFn(ctx, topic.Status.ProjectID)
+	if err != nil {
+		logging.FromContext(ctx).Desugar().Error("Failed to create Pub/Sub client", zap.Error(err))
+		return err
+	}
+	defer client.Close()
+
+	t := client.Topic(topic.Status.ProjectID)
+	exists, err := t.Exists(ctx)
+	if err != nil {
+		logging.FromContext(ctx).Desugar().Error("Failed to verify Pub/Sub topic exists", zap.Error(err))
+		return err
+	}
+
+	if !exists {
+		if topic.Spec.PropagationPolicy == v1alpha1.TopicPolicyNoCreateNoDelete {
+			logging.FromContext(ctx).Desugar().Error("Topic does not exist")
+			return fmt.Errorf("Topic %q does not exist", topic.Spec.Topic)
```

Comment:
I'd expand the error message to clarify why the topic not existing is an unrecoverable error: `Topic %q does not exist and the topic policy doesn't allow creation`

Reply:
Done! thanks!